zerver/lib/message.py

```diff
@@ -1432,3 +1432,14 @@ def set_visibility_policy_possible(user_profile: UserProfile, message: Message)
 def remove_single_newlines(content: str) -> str:
     content = content.strip("\n")
     return re.sub(r"(?<!\n)\n(?![\n0-9*-])", " ", content)
+
+
+def get_recipient_topic_by_narrow_operand_access_unchecked(
+    realm: Realm, message_id: int
+) -> Optional[Message]:
+    """This is required to get the stream and topic names from the operand
+    of `with` operator to prepare a response that user can access
+    and not send it out to unauthorized recipients.
+    """
+    message = Message.objects.filter(realm=realm, id=message_id).first()
+    return message
```

Comment:
Why aren't you using `access_message_by_id`? It's almost always a security bug to have any behavior depend on the state of the database without checking access.

Reply:
That other code path has a very explicit comment about what it is doing with the data fetched this way that is safe:
```
        # Note: It is okay here to not check access to channel                                          
        # because we are only using the channel ID to exclude data,                                     
        # not to include results.                                                                       
        channel = get_channel_from_narrow_access_unchecked(narrow, user_profile.realm)                  
```

The code path that you're adding this to does not -- if a message was moved to a private channel that you can't access, you shouldn't be able to access the new location's stream/topic.

If `maybe_user_profile` is `None`, this this request is on behalf of a public access option request; in that case, we should be able to check if the message should be available correctly; I've not tried to look at the logic here, but I think that would be a good corner case to have an explicit test for.

Reply:
oh ya, that makes sense! I'll update it.

Reply:
updated!